src/idom/server/utils.py

```diff
@@ -46,13 +49,37 @@ def run(
     host = host
     port = port or find_available_port(host)
 
-    logger.info(f"Running with {type(app).__name__!r} at http://{host}:{port}")
+    app_cls = type(app)
+    logger.info(
+        f"Running with {app_cls.__module__}.{app_cls.__name__} at http://{host}:{port}"
+    )
 
     asyncio.get_event_loop().run_until_complete(
         implementation.serve_development_app(app, host, port)
     )
 
 
+def safe_client_build_dir_path(path: str) -> Path:
+    """Prevent path traversal out of :data:`CLIENT_BUILD_DIR`"""
+    start, _, end = (path[:-1] if path.endswith("/") else path).rpartition("/")
+    file = end or start
+    final_path = traversal_safe_path(CLIENT_BUILD_DIR, file)
+    return final_path if final_path.is_file() else (CLIENT_BUILD_DIR / "index.html")
+
+
+def safe_web_modules_dir_path(path: str) -> Path:
+    """Prevent path traversal out of :data:`idom.config.IDOM_WEB_MODULES_DIR`"""
+    return traversal_safe_path(IDOM_WEB_MODULES_DIR.current, *path.split("/"))
+
+
+def traversal_safe_path(root: Path, *unsafe_parts: str | Path) -> Path:
+    """Sanitize user given path using ``werkzeug.security.safe_join``"""
+    path = safe_join(str(root.resolve()), *unsafe_parts)  # type: ignore
```

Comment:
Seems like [`safe_join` is posix specific](https://github.com/pallets/werkzeug/blob/c7ae2fea4fb229ffd71187c2b665874c91b96277/src/werkzeug/security.py#L110). I would instead just utilize our [source from `django_idom`](https://github.com/idom-team/django-idom/blob/f88a7105f6c92aafb6f69f37524e20de5fed7c40/src/django_idom/http/views.py#L17) for safe path traversal, which is OS-independent.

Reply:
Thanks for catching that. I had a solution I was using before, but I was just hesitant to implement my own. Will fix that.

Reply:
As long as you rely on pathlib `Path.absolute()` with `os.path.commonpath()` it'll be directly traversal-safe. They're fairly battle-tested.